presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.type;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.util.Objects;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+
+public final class CharType
+        extends AbstractVariableWidthType
+{
+    public static final int MAX_LENGTH = 65_536;
+
+    public static CharType createCharType(long length)
+    {
+        return new CharType(length);
+    }
+
+    private final int length;
+
+    private CharType(long length)
+    {
+        super(
+                new TypeSignature(
+                        StandardTypes.CHAR,
+                        singletonList(TypeSignatureParameter.of(length))),
+                Slice.class);
+
+        if (length < 1 || length > MAX_LENGTH) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("CHAR length scale must be in range [1, %s]", MAX_LENGTH));
+        }
+        this.length = (int) length;
+    }
+
+    public int getLength()
+    {
+        return length;
+    }
+
+    @Override
+    public boolean isComparable()
+    {
+        return true;
+    }
+
+    @Override
+    public boolean isOrderable()
+    {
+        return true;
+    }
+
+    @Override
+    public Object getObjectValue(ConnectorSession session, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        return block.getSlice(position, 0, block.getLength(position)).toStringUtf8();
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        int leftLength = leftBlock.getLength(leftPosition);
+        int rightLength = rightBlock.getLength(rightPosition);
+        if (leftLength != rightLength) {
+            return false;
+        }
+        return leftBlock.equals(leftPosition, 0, rightBlock, rightPosition, 0, leftLength);
```

Comment:
I agree that it shouldn't work like this, but unfortunately it does. @haozhun and I have been trying to clean this up in https://github.com/prestodb/presto/pull/5781 and https://github.com/prestodb/presto/pull/5778, but after that goes in more work needs to be done to remove "type only" coercions.

As I understand it, type only coercions have the following contract:
1) Implementation of the CAST must be the identity function.
2) All code in Presto, after analysis, must assume that type only coercions may be implicitly present.

I see a few ways forward here:
1) Add a flag like `experimental.enable-char-type` which would default to `false`, document all the known issues with `char(x)` and merge it, with that flag so that people can try it out.
2) Move all the `char(x)` features that don't work currently into a separate PR (coercions, `char(x)`/`char(y)` comparison...) and just merge the basics.
3) Fix the issues in our type system to support `char(x)` and then merge this PR afterwards.

Let me know which of those would work best for you.

Moving `equalTo`, `hash`, and `compareTo` on `Block` into the `FunctionRegistry` is something we've been talking about for a while. That's probably the best option, but one thing we should test first is whether those functions are actually necessary, or if the `EQUAL` operator could just be used by invoking the `MethodHandle` in a loop.